### PR TITLE
Fix creation/alteration of scalar constraints used in containers

### LINKED ISSUE #1370: InternalServerError: cannot alter type
- EdgeDB Version: EdgeDB 1.0-alpha.2+g16f75ece5.d20200122
- OS Version: Windos 10 Educational, Version: 1909, Build: 18363.778

Steps to Reproduce:
```
START TRANSACTION;
CREATE MIGRATION m2 TO {
    module default {
        type User {
            required property scopes -> array<scope>;
        }
        scalar type scope extending str {
            constraint one_of (
                '1', '2'
            );
        }
    }
};
COMMIT MIGRATION m2;
COMMIT;
```

Error log:
```
InternalServerError: cannot alter type ""edgedb_e1d3ab02-7c59-11ea-9cc7-f7e49d2fb386"."87c9d94a-87f0-11ea-b4ba-db2bb9dbfd2b_domain"" because column "6716f2a1-87ef-11ea-bb56-fd1198a88276.scopes" uses it
```

May be related to this issue #1369.

### PR DESCRIPTION
**Pull request by msullivan**
Migrations that created a scalar type and a property that used it
inside a container were broken for a simple reason, which is that the
constraint creation would get priority sorted after the column
creation. Tweak things around to avoid that.

However, adding a constraint to an existing scalar type that was used
in a property inside of a container type
(like `property x -> array<positive_int>`) was broken in a more
fundamental way, because pgsql disallows doing that.

We can work around this in a laborious way:
 1. Find all of the properties whose type is a container type
    that contains this scalar. (Possibly transitively.)
 2. Change the type of all offending properties to an equivalent type
    that does not reference this scalar. This may require creating
    new types.
 3. Add the constraint.
 4. Restore the type of all offending properties. If existing data
    violates the new constraint, we will fail here. Delete any
    temporarily created types.

This is a lot more code than the importance of the particular case
really ought to justify, but we live in a fallen world.

The SET TYPEs work by generating automatic conversion code using
casts. All of this infrastructure existed already but getting it
to work for the tests I wrote required fixing a couple cast bugs
and a bug in the compilation of aggregate functions used directly
as a query top-level.

Fixes #1370.

### CODE REVIEW

File: edb/pgsql/compiler/relgen.py [hunk 1]
```diff
@@ -2556,7 +2556,7 @@ def process_set_as_agg_expr(
                 ctx=xctx)
 
     if wrapper:
-        stmt = wrapper
+        ctx.rel = stmt = wrapper
```
Comment (elprans) [inline edb/pgsql/compiler/relgen.py:2559]: This seems fishy.  `wrapper` may come from a `ctx.subrel()` so we're doing some weird inversion here.  Why is this needed?

File: edb/pgsql/delta.py [hunk 2]
```diff
@@ -1718,8 +1863,18 @@ def apply(
         orig_schema = schema
         schema = s_scalars.AlterScalarType.apply(self, schema, context)
         new_scalar = self.scls
+
+        used_props = []
+        if list(self.get_subcommands(type=s_constr.CreateConstraint)):
+            used_props = self._get_composite_refs(schema, context)
+            schema = self._undo_everything(schema, context, used_props)
```
Comment (elprans) [inline edb/pgsql/delta.py:1870]: Maybe add `if used_props:` so this reads cleaner.

File: edb/pgsql/compiler/relgen.py (hunk 1 above)
Comment (msullivan) [reply edb/pgsql/compiler/relgen.py]: When a aggregate function with an initial_val was used at the top-level of a query (without a SelectStmt wrapping it, which happens for fragments like the USING clauses of SET TYPE), the actual results were in one of the rvars in the returned SetRVars but they weren't actually making it into the result rel.

This is because in those cases, we were totally ignoring the input stmt in favor of creating a new statement, which was dubious. This fix was even more dodgy, and I've pushed a cleanup.

(Though it also feels off to me that `_process_toplevel_query` doesn't do anything with the rvars in the common case. But trying to `_include_rvars` there causes us to try to generate an infinite amount of SQL)

**Review by elprans — APPROVED** [review]
:+1:
